src/lib/coreaudio/CoreAudioStream.cxx

```diff
@@ -0,0 +1,226 @@
+		if (original_virtual_fmt.mFormatID) {
+			FormatDebug(macos_output_domain, "Restoring original virtual format for stream 0x%04x. (%s)", (uint)stream_id, StreamDescriptionToString(original_virtual_fmt).c_str());
+			SetVirtualFormat(original_virtual_fmt);
+		}
+		if (original_physical_fmt.mFormatID) {
+			FormatDebug(macos_output_domain, "Restoring original physical format for stream 0x%04x. (%s)", (uint)stream_id, StreamDescriptionToString(original_physical_fmt).c_str());
+			SetPhysicalFormat(original_physical_fmt);
+		}
+	}
+	catch (...) {
+		// Ignore any exceptions that might be thrown, as the stream is anyways closed.
+	}
+	original_virtual_fmt.mFormatID  = 0;
+	original_physical_fmt.mFormatID = 0;
+	FormatDebug(macos_output_domain, "Closed stream 0x%04x.", (uint)stream_id);
+	stream_id = 0;
+}
+
+AudioStreamBasicDescription
+CoreAudioStream::GetVirtualFormat() {
+	if (!stream_id)
+		throw std::runtime_error("Invalid stream ID.");
+	AudioStreamBasicDescription desc;
+	UInt32 size = sizeof(desc);
+
+	AudioObjectPropertyAddress aopa;
+	aopa.mScope = kAudioObjectPropertyScopeGlobal;
+	aopa.mElement = kAudioObjectPropertyElementMaster;
+	aopa.mSelector = kAudioStreamPropertyVirtualFormat;
+
+	OSStatus err = AudioObjectGetPropertyData(stream_id, &aopa, 0, NULL, &size, &desc);
+	if (err != noErr)
+		throw FormatRuntimeError("Unable to retrieve virtual format for stream 0x%04x.", (uint)stream_id);
+	return desc;
+}
+
+void
+CoreAudioStream::SetVirtualFormat(AudioStreamBasicDescription desc) {
+	if (!stream_id)
+		return;
+
+	AudioObjectPropertyAddress aopa;
+	aopa.mScope = kAudioObjectPropertyScopeGlobal;
+	aopa.mElement = kAudioObjectPropertyElementMaster;
+	aopa.mSelector = kAudioStreamPropertyVirtualFormat;
+
+	UInt32 property_size = sizeof(AudioStreamBasicDescription);
+	OSStatus err = AudioObjectSetPropertyData(stream_id, &aopa, 0, NULL, property_size, &desc);
+	if (err != noErr)
+		throw FormatRuntimeError("Unable to set virtual format for stream 0x%04x. Error = %s", (uint)stream_id, GetError(err));
+}
+
+AudioStreamBasicDescription
+CoreAudioStream::GetPhysicalFormat() {
+	if (!stream_id)
+		throw std::runtime_error("Invalid stream ID.");
+
+	AudioStreamBasicDescription desc;
+	UInt32 size = sizeof(desc);
+
+	AudioObjectPropertyAddress aopa;
+	aopa.mScope = kAudioObjectPropertyScopeGlobal;
+	aopa.mElement = kAudioObjectPropertyElementMaster;
+	aopa.mSelector = kAudioStreamPropertyPhysicalFormat;
+
+	OSStatus err = AudioObjectGetPropertyData(stream_id, &aopa, 0, NULL, &size, &desc);
+	if (err != noErr)
+		throw FormatRuntimeError("Unable to retrieve physical format for stream 0x%04x.", (uint)stream_id);
+	return desc;
+}
+
+void
+CoreAudioStream::SetPhysicalFormat(AudioStreamBasicDescription desc) {
+	if (!stream_id)
+		return;
+
+	AudioObjectPropertyAddress aopa;
+	aopa.mScope = kAudioObjectPropertyScopeGlobal;
+	aopa.mElement = kAudioObjectPropertyElementMaster;
+	aopa.mSelector = kAudioStreamPropertyPhysicalFormat;
```

Comment:
please keep in mind that setting physical format is time consuming. System/DACs may skipped a few samples if you set it. That's part of the reason I deleted relevant code for cmus https://github.com/cmus/cmus/pull/456/commits/995e0673522e191628e9319be6d553f116c21a48 a while ago. 

If you want to have this feature, please make it an opt-in feature.

Reply:
You can play some music with any music player app you have, and go to "Audio Midi Setup" app, and change your DAC setting for sample rate / bit depth. You can clearly hear the hiccup when you change it.

Reply:
Well, this is clear since you are then changing the format while music is playing. This will never be done by this plugin though. Sample rate / bit depth is only changed in beginning of playback of a new song which has a different format than the previous one.